File: go/internal/renderer/camera.go
package renderer

import (
	"math"
	"math/rand"

	"github.com/nya3jp/raytracing/go/internal/geom"
)

type Camera struct {
	origin          geom.Vec
	lowerLeftCorner geom.Vec
	horizontal      geom.Vec
	vertical        geom.Vec
	u               geom.Vec
	v               geom.Vec
	lensRadius      float64
}

func NewCamera(lookFrom, lookAt, up geom.Vec, fov, aspectRatio, aperture, focusDist float64) *Camera {
	viewportHeight := 2 * math.Tan(fov/2)
	viewportWidth := viewportHeight * aspectRatio

	w := lookAt.Sub(lookFrom).Unit()
	u := w.Cross(up).Unit()
	v := u.Cross(w)

	horizontal := u.Mul(focusDist * viewportWidth)
	vertical := v.Mul(focusDist * viewportHeight)
	lowerLeftCorner := lookFrom.Add(w.Mul(focusDist)).Sub(horizontal.Div(2)).Sub(vertical.Div(2))
	return &Camera{
		origin:          lookFrom,
		lowerLeftCorner: lowerLeftCorner,
		horizontal:      horizontal,
		vertical:        vertical,
		u:               u,
		v:               v,
		lensRadius:      aperture / 2,
	}
}

func (c *Camera) Ray(s, t float64, random *rand.Rand) *geom.Ray {
	lens := geom.RandomVecInUnitDisc(random).Mul(c.lensRadius)
	blur := c.u.Mul(lens.X).Add(c.v.Mul(lens.Y))
	origin := c.origin.Add(blur)
	dir := c.lowerLeftCorner.Add(c.horizontal.Mul(s)).Add(c.vertical.Mul(t)).Sub(origin)
	return geom.NewRay(origin, dir)
}

File: rust/src/camera.rs
use crate::geom::{Ray, Vec3};
use crate::rng::Rng;

#[derive(Debug)]
pub struct Camera {
    origin: Vec3,
    horizontal: Vec3,
    vertical: Vec3,
    lower_left_corner: Vec3,
    u: Vec3,
    v: Vec3,
    lens_radius: f64,
}

impl Camera {
    pub fn new(
        origin: Vec3,
        look_at: Vec3,
        fov: f64,
        aspect_ratio: f64,
        aperture: f64,
        focus_dist: f64,
    ) -> Camera {
        let viewport_height = 2.0 * (fov / 2.0).atan();
        let viewport_width = viewport_height * aspect_ratio;

        let w = (look_at - origin).unit();
        let up = Vec3::new(0.0, 1.0, 0.0);
        let u = w.cross(up).unit();
        let v = u.cross(w);

        let horizontal = u * (focus_dist * viewport_width);
        let vertical = v * (focus_dist * viewport_height);
        let lower_left_corner = origin + w * focus_dist - horizontal / 2.0 - vertical / 2.0;
        Camera {
            origin,
            horizontal,
            vertical,
            lower_left_corner,
            u,
            v,
            lens_radius: aperture / 2.0,
        }
    }

    pub fn ray(&self, u: f64, v: f64, rng: &mut Rng) -> Ray {
        let lens = Vec3::random_in_unit_disc(rng) * self.lens_radius;
        let blur = self.u * lens.x + self.v * lens.y;
        let origin = self.origin + blur;
        let target = self.lower_left_corner + self.horizontal * u + self.vertical * v;
        Ray::new(origin, target - origin)
    }
}
